### PR TITLE
refactor(server): Refactor `AddOrFind()`

### PR DESCRIPTION
**Pull request by chakaz**
This actually fixes a bug where we did not handle tiering upon expiration if found via `AddOrFind2()`.

Related to #2252

### CODE REVIEW

**Comment by chakaz** [comment]
@adiholden please LMK if there's anything further you wanted to see in this small refactor

**Comment by adiholden** [comment]
@chakaz I thought you were going to simplify AddOrFind2

File: src/server/table.cc
Comment (adiholden) [reply src/server/table.cc:37]: what do you mean PreUpdate() will no longer do anything ?

Comment (chakaz) [reply src/server/table.cc:37]: I mean that it will not remove object memory just for `PostUpdate()` to add it back again.
I believe that `update_value_amount` is introduced to learn how much memory is temporarily removed, or to find "leaks" (that we remove and not add). Both will be irrelevant soon, so I think that there's no point in keeping this counter.

**Comment by chakaz** [comment]
I *really* like your proposal. When I went about refactoring this function I did not think that this optimization can be removed, but I was just thinking inside the box.
The code now is *much* simpler!

File: src/server/db_slice.cc
Comment (adiholden) [inline src/server/db_slice.cc:506]: Why do you think we should not update the stats here? maybe not the miss stats but bumpup and hit I believe we should , no?

File: src/server/db_slice.cc
Comment (adiholden) [inline src/server/db_slice.cc:559]: Note that in my suggestion I wrote about Adding a new function to dash table that on Insert will not try to find if the value exists. As in the flow now you do 2 find operations in dash table if this is a new entry

File: src/server/db_slice.cc
Comment (chakaz) [reply src/server/db_slice.cc:506]: The reason is that [this](https://github.com/dragonflydb/dragonfly/blob/995f3cd8aa9caf616ebe392ca3f795daffeab7ba/src/server/multi_test.cc#L135) test failed, and I do think it's weird for `SET` to apply hits, but it's a matter of decision.
Whatever decision we make, it needs to be consistent, as it turns out that before this change the stats were different based on whether there were callbacks :facepalm:

File: src/server/db_slice.cc
Comment (chakaz) [reply src/server/db_slice.cc]: Right, I forgot about that. Done now.

File: src/server/db_slice.cc
Comment (adiholden) [reply src/server/db_slice.cc:506]: Ok. Maybe we want to update hits/misses only in case we call FindReadOnly.
Cause right now if we call set we dont update the countes and if we call append which invokes FindMutable we will update this stats.
Anyway I believe we do what to update the bumpups stats even if we dont update the hit/miss

Comment (chakaz) [reply src/server/db_slice.cc:505]: Makes sense. Done.

File: src/server/db_slice.cc
Comment (adiholden) [inline src/server/db_slice.cc:574]: I dont think you should call PreUpdate here as PreUpdate should be call before we update the value of an entry. The only common functionality here is the SetVerstion and you can call it explicitly

Comment (chakaz) [reply src/server/db_slice.cc:574]: I find this more future compatible, but any changes we may make in `PreUpdate()`.
But since you prefer that way, I won't argue. Done. I'd be happy to revert if you change your mind though :)

File: src/server/db_slice.cc
Comment (adiholden) [inline src/server/db_slice.cc:335]: I believe this also should be kDontUpdateStats right?

File: src/server/db_slice.cc
Comment (adiholden) [inline src/server/db_slice.cc:424]: also here kDontUpdateStats? Only FindReadOnly should have the kUpdateStats I think

File: src/server/db_slice.h
Comment (adiholden) [inline src/server/db_slice.h:417]: nit: as this is now only relevant to hit/miss stats maybe call it UpdateCacheStats

File: src/server/db_slice.cc
Comment (chakaz) [reply src/server/db_slice.cc:335]: `Find()` is about to be removed (https://github.com/dragonflydb/dragonfly/pull/2290), so it doesn't matter greatly, but sure :)

File: src/server/db_slice.cc
Comment (adiholden) [reply src/server/db_slice.cc:574]: note: you dont need the PreUpdateMode enum now.

File: src/server/db_slice.cc
Comment (chakaz) [reply src/server/db_slice.cc:424]: `FindExt()` should be changed soon, but it can be used for both read-only and modify operations. Like `Find()` was before I added `FindReadOnly()`.
Since it's causing some tests to fail (and changes existing behavior), let's leave it as is for now and revisit in the other PR...

File: src/server/db_slice.h
Comment (chakaz) [reply src/server/db_slice.h:417]: Lovely!

**Comment by adiholden** [comment]
Before this PR is getting merged @romange I want to get your approval on the functional changes of this PR which simplifies the logic of AddORFind and fixes some bugs in the exising flow.
The changes are:
1. We had an optimization in AddOrFind that in case there are no registered callbacks we Call dash table insert. If the insert returns an existing entry and the existing entry is expired we reset the PrimeValue and return the iterator. Now we always first FindInternal function that first checks if the entry exists if so we expire the entry and add the entry again with dash table insert.
2. We now only update hit/miss stats for read commands. i.e set x y will not incr the hit counter if x is found in table.

@chakaz Thank you bearing with me :)

**Comment by chakaz** [comment]
> 2. We now only update hit/miss stats for read commands. i.e set x y will not incr the hit counter if x is found in table.

To clarify, our previous behavior was bad: we used to update these counters only if replication was not enabled (i.e. no callbacks).
I do think that we shouldn't update hit/miss on modify operations like `SET`, but even if we do, we'll make it consistent.

> @chakaz Thank you bearing with me :)

Thank you! The code now, imho, is much better!

**Comment by romange** [comment]
@adiholden @chakaz  LGTM! Thank you for simplifying the code there and finding all these bugs! :)
Great work! 💪🏼

**Comment by chakaz** [comment]
Sure thing, but can I haz approval? :pray:

**Review by adiholden — APPROVED** [review]

**Pull request merged**
